### PR TITLE
odb: Modifies abstract LEF writer to work on ostreams 

### PR DESCRIPTION
**Pull request by QuantamHD**
Moves file output mechanism to ostreams instead of FILE*. I want an ostream so I can run unit tests without having to rely on the filesystem. Also it's just nice to modernize this interface

### CODE REVIEW

File: src/odb/src/lefout/lefout.cpp [hunk 1]
```diff
@@ -897,57 +908,59 @@ void lefout::writeVia(dbTechVia* via)
...
-            lefdist(P.getXTopEnclosure()),
-            lefdist(P.getYTopEnclosure()));
+    fmt::print(
+        _out, " + LAYERS {} {} {} ", bot.c_str(), cut.c_str(), top.c_str());
+    fmt::print(_out,
+               " + CUTSPACING {:.11g} {:.11g} ",
+               lefdist(P.getXCutSpacing()),
+               lefdist(P.getYCutSpacing()));
+    fmt::print(_out,
+               " + ENCLOSURE {:.11g} {:.11g} {:.11g} {:.11g} ",
```
Comment (maliberty) [inline src/odb/src/lefout/lefout.cpp:930]: While not an issue with the translation, it seems there should be new lines in these statements.

File: src/odb/src/swig/common/swig_common.cpp [hunk 2]
```diff
@@ -127,22 +127,28 @@ int write_def(odb::dbBlock* block,
 
 int write_lef(odb::dbLib* lib, const char* path)
 {
+  std::ofstream os(path);
```
Comment (maliberty) [inline src/odb/src/swig/common/swig_common.cpp:130]: Check if successful

File: src/odb/src/swig/common/swig_common.cpp [hunk 3]
```diff
@@ -127,22 +127,28 @@ int write_def(odb::dbBlock* block,
...
-  odb::lefout writer(logger);
-  return writer.writeTechAndLib(lib, path);
+  odb::lefout writer(logger, os);
+  writer.writeTechAndLib(lib);
+  return true;
 }
 
 int write_tech_lef(odb::dbTech* tech, const char* path)
 {
+  std::ofstream os(path);
```
Comment (maliberty) [inline src/odb/src/swig/common/swig_common.cpp:139]: Check if successful

File: src/odb/src/swig/common/swig_common.cpp [hunk 4]
```diff
@@ -127,22 +127,28 @@ int write_def(odb::dbBlock* block,
...
   utl::Logger* logger = new utl::Logger(NULL);
-  odb::lefout writer(logger);
-  return writer.writeTech(tech, path);
+  odb::lefout writer(logger, os);
+  writer.writeTech(tech);
+  return true;
 }
 int write_macro_lef(odb::dbLib* lib, const char* path)
 {
+  std::ofstream os(path);
```
Comment (maliberty) [inline src/odb/src/swig/common/swig_common.cpp:147]: Check if successful

**Review by maliberty — CHANGES_REQUESTED** [review]

**Review by maliberty — APPROVED** [review]

File: src/OpenRoad.cc [hunk 5]
```diff
@@ -390,17 +390,26 @@ void OpenRoad::writeLef(const char* filename)
       if (cnt > 0) {
         name += "_" + std::to_string(cnt);
         std::ofstream os(name);
+        if (!os.is_open()) {
```
Comment (oharboe) [inline src/OpenRoad.cc:393]: Exceptions are enabled when reading. The advantage is that *all* IO operations are checked, including each individual read/write operation, so if one runs out of e.g. disk space, an exception is thrown.

https://github.com/The-OpenROAD-Project/OpenROAD/blob/7586d03811ca43cc2615077afb0f4821c81f82e2/src/OpenRoad.cc#L426

The code below has is manually adding the exceptions.

**Review by oharboe — CHANGES_REQUESTED** [review]
enable exceptions in ofstream to get error checking everywhere, as is done for read:

https://github.com/The-OpenROAD-Project/OpenROAD/blob/7586d03811ca43cc2615077afb0f4821c81f82e2/src/OpenRoad.cc#L426

File: src/OpenRoad.cc (hunk 5 above)
Comment (QuantamHD) [reply src/OpenRoad.cc:393]: @maliberty Do we correctly catch these exceptions in TCL land?

Comment (maliberty) [reply src/OpenRoad.cc:393]: https://github.com/The-OpenROAD-Project/OpenROAD/blob/7586d03811ca43cc2615077afb0f4821c81f82e2/src/Exception.i#L28

should catch them but you can always verify

**Review by maliberty — APPROVED** [review]

**Pull request merged**